### PR TITLE
{containerapp} support connected env command group

### PR DESCRIPTION
**Pull request by Greedygre**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command

az containerapp connected-env show/list/delete/create

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
containerapp

File: src/containerapp/azext_containerapp/_params.py [hunk 1]
```diff
@@ -519,3 +519,15 @@ def load_arguments(self, _):
...
         c.argument('environment_type', arg_type=get_enum_type(["managed", "connected"]), help="Type of environment.", is_preview=True)
+
+    with self.argument_context('containerapp connected-env') as c:
+        c.argument('name', name_type, help='Name of the Container Apps connected environment.')
+        c.argument('resource_group_name', arg_type=resource_group_name_type)
+        c.argument('tags', arg_type=tags_type)
+        c.argument('custom_location', help="Resource ID of custom location. List using 'az customlocation list'.", validator=validate_custom_location_name_or_id)
+
+    with self.argument_context('containerapp connected-env', arg_group='Dapr') as c:
+        c.argument('dapr_ai_connection_string', options_list=['--dapr-ai-connection-string', '--dapr-connection'], help='Connection string used for Dapr application insights.')
```
[the remark this answers is not in the document]
Comment (Greedygre) [reply src/containerapp/azext_containerapp/_params.py:530]: The `--dapr-ai-connection-string` is too long, need a shorter name.

File: src/containerapp/azext_containerapp/_params.py [hunk 2]
```diff
@@ -519,3 +519,15 @@ def load_arguments(self, _):
...
+
+    with self.argument_context('containerapp connected-env', arg_group='Network') as c:
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_params.py:532]: no need to add argument group when there is only one parameter

File: src/containerapp/azext_containerapp/_params.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Juliehzl) [reply src/containerapp/azext_containerapp/_params.py:530]: should we only keep one for current options? Typically for paratemerts with two options, one is meaningful and another one is -x

File: src/containerapp/azext_containerapp/_params.py [hunk 3]
```diff
@@ -519,3 +519,15 @@ def load_arguments(self, _):
...
+        c.argument('static_ip', help='Static IP value.')
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_params.py:533]: reference swagger description

File: src/containerapp/azext_containerapp/_utils.py [hunk 4]
```diff
@@ -2083,3 +2087,53 @@ def parse_oryx_mariner_tag(tag: str) -> OryxMarinerRunImgTagProperty:
     else:
         tag_obj = None
     return tag_obj
+
+
+def get_custom_location(cmd, custom_location_id):
+    parsed_custom_loc = parse_resource_id(custom_location_id)
+    subscription_id = parsed_custom_loc.get("subscription")
+    custom_loc_name = parsed_custom_loc["name"]
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_utils.py:2095]: why use different way to get value?

File: src/containerapp/azext_containerapp/_utils.py [hunk 5]
```diff
@@ -2083,3 +2087,53 @@ def parse_oryx_mariner_tag(tag: str) -> OryxMarinerRunImgTagProperty:
...
+def validate_custom_location(cmd, custom_location=None):
+    if not is_valid_resource_id(custom_location):
+        raise ValidationError('{} is not a valid Azure resource ID.'.format(custom_location))
+
+    r = get_custom_location(cmd=cmd, custom_location_id=custom_location)
+    if r is None:
+        raise ResourceNotFoundError("Cannot find custom location with custom location ID {}".format(custom_location))
+
+    # check extension type
+    check_extension_type = False
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/_utils.py:2131]: should the var name with extension_existing?

File: src/containerapp/azext_containerapp/commands.py [hunk 6]
```diff
@@ -231,3 +231,9 @@ def load_command_table(self, _):
         g.custom_command('list', 'patch_list')
         g.custom_command('apply', 'patch_apply')
         g.custom_command('interactive', 'patch_interactive')
+
+    with self.command_group('containerapp connected-env', is_preview=True) as g:
+        g.custom_show_command('show', 'show_connected_environment')
+        g.custom_command('list', 'list_connected_environments')
+        g.custom_command('create', 'create_connected_environment', supports_no_wait=True, exception_handler=ex_handler_factory())
+        g.custom_command('delete', 'delete_connected_environment', supports_no_wait=True, confirmation=True, exception_handler=ex_handler_factory())
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/commands.py:239]: what about update?

Comment (Greedygre) [reply src/containerapp/azext_containerapp/commands.py:239]: Will add update next PR.

File: src/containerapp/azext_containerapp/_params.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Greedygre) [reply src/containerapp/azext_containerapp/_params.py:530]: I will keep options_list=['--dapr-ai-connection-string', '-d']

**Comment by Greedygre** [comment]
Hi @zhoxing-ms @wangzelin007 
The (Verify Ref Docs)  failed, could you help to handle? Thanks.

File: src/containerapp/azext_containerapp/tests/latest/test_containerapp_preview_scenario.py [hunk 7]
```diff
@@ -57,38 +44,50 @@ def test_containerapp_preview_environment_type(self, resource_group):
...
-            .format(location=TEST_LOCATION, custom_location_id=custom_location_id)
-        write_test_file(file, env_payload)
-        self.cmd(f'az rest --method put --uri "{connected_env_resource_id}?api-version=2022-06-01-preview" --body "@{file}"')
-        containerapp_env = self.cmd(f'az rest --method get --uri "{connected_env_resource_id}?api-version=2022-06-01-preview"').get_output_in_json()
-        while containerapp_env["properties"]["provisioningState"].lower() != "succeeded":
-            time.sleep(5)
-            containerapp_env = self.cmd(
-                f'az rest --method get --uri "{connected_env_resource_id}?api-version=2022-06-01-preview"').get_output_in_json()
+        custom_location_id = f"/subscriptions/{sub_id}/resourceGroups/{resource_group}/providers/Microsoft.ExtendedLocation/customLocations/{custom_location_name}"
+        self.cmd(f'containerapp connected-env create -g {resource_group} --name {connected_env_name} --custom-location {custom_location_name} --static-ip {static_ip} -d "InstrumentationKey=TestInstrumentationKey;IngestionEndpoint=https://ingestion.com/;LiveEndpoint=https://abc.com/" -l {TEST_LOCATION}', checks=[
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/tests/latest/test_containerapp_preview_scenario.py:56]: seperate to connected_env_scenario_test

Comment (Greedygre) [reply src/containerapp/azext_containerapp/tests/latest/test_containerapp_preview_scenario.py:56]: I will seperate to connected_env_scenario_test in next PR.

**Review by Juliehzl — APPROVED** [review]

**Comment by Greedygre** [comment]
Hi @zhoxing-ms 
Can you help to review and merge this PR, thanks.

File: src/containerapp/setup.py [hunk 8]
```diff
@@ -47,7 +47,9 @@
 ]
 
 # TODO: Add any additional SDK dependencies here
-DEPENDENCIES = ['pycomposefile>=0.0.29', 'docker']
+DEPENDENCIES = ['pycomposefile>=0.0.29',
+                'docker',
+                'azure-mgmt-kubernetesconfiguration==2.0.0']
```
Comment (zhoxing-ms) [inline src/containerapp/setup.py:52]: @jiasli May I ask the azure-xxx SDK related dependencies cannot be defined in `setup.py`, right?

**Comment by Greedygre** [comment]
Hi @zhoxing-ms 
Comment addressed, could you please help to review and approve? Thanks.

**Comment by zhoxing-ms** [comment]
`vendored_sdks/kubernetesconfiguration` --> `vendored_sdks/kubernetes_configuration`

Suggest making the naming of folder for vendored sdk more readable

**Comment by Greedygre** [comment]
> `vendored_sdks/kubernetesconfiguration` --> `vendored_sdks/kubernetes_configuration`
> 
> Suggest making the naming of folder for vendored sdk more readable

Hi, this directory name is consisting with the azure-mgmt-kubernetesconfiguration.

**Pull request merged**